lib/signer.ex

```diff
@@ -40,42 +40,54 @@ defmodule AntikytheraAws.Signer do
     - `params` must be decoded query strings, in a list of tuple-2
         - To convert from `Antikythera.Http.QueryParams` (which is `Croma.SubtypeOfMap`), just use `Map.to_list/1`
 
-    Note: "content-type" and "x-amz-content-sha256" headers are required for S3, but not generally.
+    Note: "content-type" header is required for S3, but not generally.
     """
     defun prepare_headers(%Creds{access_key_id: aki, secret_access_key: sak, security_token: st} = _creds,
                           region  :: v[String.t],
                           service :: v[String.t],
                           method  :: v[Method.t],
                           path    :: v[UPath.t],
-                          payload :: v[String.t],
+                          payload :: v[binary],
                           headers :: v[Headers.t],
                           params  :: [{String.t, String.t}]) :: Headers.t do
       if String.contains?(path, "?"), do: raise("`path` must not include query string part!")
       downcased_headers = Map.new(headers, fn {key, val} -> {String.downcase(key), val} end)
       if !Map.has_key?(downcased_headers, "host"), do: raise("'host' header is required!")
-      {amz_date, headers_with_date} = case downcased_headers do
-        %{"x-amz-date" => ad} -> {ad, downcased_headers}
-        %{"date" => date}     ->
-          ad = Time.from_http_date(date) |> Croma.Result.get!() |> Time.to_iso_basic()
-          {ad, downcased_headers}
-        without_amz_date      ->
-          ad = Time.to_iso_basic(Time.now())
-          {ad, Map.put(without_amz_date, "x-amz-date", ad)}
-      end
-      sign_ready_headers = if st, do: Map.put(headers_with_date, "x-amz-security-token", st), else: headers_with_date
+      {amz_date, headers_with_date} = gen_amz_date_and_put_date(downcased_headers)
+      headers_might_with_st = if st, do: Map.put(headers_with_date, "x-amz-security-token", st), else: headers_with_date
+      payload_sha256        = hex_sha256(payload)
+      sign_ready_headers    = put_content_sha256_if_s3_request(headers_might_with_st, service, payload_sha256)
```

Comment:
We can further refactor this function into something like:
```ex
defun ... do
  if String.contains?(path, "?"), do: raise("`path` must not include query string part!")
  payload_sha256     = hex_sha256(payload)
  sign_ready_headers = canonicalize_headers(headers, st, service, payload_sha256)
  canonical_request  = make_canonical_request(sign_ready_headers, method, path, payload_sha256, params)
  signature          = compute_signature(aki, sak, amz_date, region, service, canonical_request)
  Map.put(sign_ready_headers, "authorization", signature)
end
```
Please do adjust the order of functions in this module as you think appropriate.

Reply:
fixed in 975a684, 77aecfd